### DISCUSSION TITLE
Discussion: ValueObservable / BehaviorObservable

### DISCUSSION THREAD

**Discussion by rozaxe** [ReactiveX/rxjs#5975] [Ideas / Feature request]
## Feature Request: ValueObservable / BehaviorObservable

**Is your feature request related to a problem? Please describe.**
RxJS works really great in the asynchronous world. However, there is plenty of times where a synchronous stream is required.

The main example is when building a web app where components must render synchronously some observable's value. A lot of times I use BehaviorSubject to do that but then I expose the mutable method "next" to the outer world.

I would like to expose only a type of **"Observable + Synchronous value"**.

**Describe the solution you'd like**
I would like a subclass of Observable, named like "ValueObservable" or "BehaviorObservable" which is a specialized Observable which has access to it's current value. Something along:

```ts
type ValueObservable<T> = Observable<T> & {
	readonly value: T
	getValue(): T
}
```

Then, BehaviorSubject would be of type `class BehaviorSubject<T> extends Subject<T> implements ValueObservable<T>`

***

Thanks to this addition we would be able to differentiate stream which already provides a value and stream who are truly asynchronous and need special care for fallback value.

In a todo app, `todos$` would be of type `ValueObservable<Todo[]>` because there is always some todos available to render while `fetchTodoFromApi$()` would still be of type `Observable<Todo[]>` because we do not have any data before the request is done.

**Describe alternatives you've considered**
I already implements this idea in the user-land package [r-value-observable](https://github.com/rozaxe/r-value-observable). But I think it would be better to have it in the core package.

**Additional context**
This idea comes from [RxDart](https://github.com/ReactiveX/rxdart) which [already implements this](https://pub.dev/documentation/rxdart/latest/rx/ValueStream-class.html) and can be used in Flutter to render synchronous value in components.

***

PS: It would be a pleasure for me to develop it and offer a PR to this package!

**Comment by ivan7237d** [comment]
I actually have a related question: when designing APIs that use observables, I use a rule of a thumb that where possible, an observable should represent a changing value rather than a stream of events. If anyone uses the same principle or knows of any articles on this subject, give a shout!

**Comment by backbone87** [comment]
given a hot observable you can just do this:
```ts
// setup
const subject = new BehaviorSubject(undefined);
a$.subscribe(subject);

// use later
const value = subject.getValue();
```

**Reply by Sawtaytoes** [reply]
`BehaviorSubject` has a `.getValue()` function?

**Reply by backbone87** [reply]
https://rxjs-dev.firebaseapp.com/api/index/class/BehaviorSubject

**Reply by mattkime** [reply]
@backbone87 I'm not entirely sure how your suggestion relates to the original proposal. As best I can tell, it doesn't guarantee `getValue` will return a value.

**Comment by nandin-borjigin** [comment]
`BehaviorSubject` is not a perfect solution as it also exposes `Observer` interface which can be used to send values to the Subject.

The original request is to have a readonly (Observable only) interface with `getValue()` capability

**Reply by voliva** [reply]
https://github.com/re-rxjs/react-rxjs already has this on their `StateObservable`s (which is published as standalone from react in https://github.com/rx-state/core), but it has a bit of a caveat.

To begin with, this only works with multicast observables... Because unicast observables could have different values. And how do you manage the subscription of that observable? Does it start as soon as it's declared? Does it start when there's one subscriber? Does it stop when all subscribers unsubscribed?

But also, what happens if you call `.getValue()` on an Observable that hasn't emitted anything yet? There are arguments that it should throw an exception, return undefined, or return a promise. And then, what happens if you call `.getValue()` but the subscription didn't even start?

React-RxJS went with:
- only `.subscribe(` starts subscription.
- subscriptions stops at refcount=0 (and restarts when a new subscriber comes in).
- .getValue() throws if the subscription is not active.
- .getValue() returns a promise if the observable didn't emit anything yet.

It went with these assumptions because they makes sense, and they also play well with React, but then the types are harder to work with:
```ts
export interface StateObservable<T> extends Observable<T> {
  getValue: () => T | Promise<T>
}

// Example...
import { state } from '@rx-state/core';

const myState$ = state(interval(1000));
myState$.subscribe(); // Start the subscription.

// After some time...
const res = myState$.getValue();
// res can be a promise or a number, depending if `interval` emitted something or not
```

We also added a different type of Observable that makes it easier, but then you have to pass in a default value: DefaultedStateObservable. The assumptions change to:

- only `.subscribe(` starts subscription.
- subscriptions stops at refcount=0 (and restarts when a new subscriber comes in).
- .getValue() returns the default value if the subscription is not active or if it didn't emit anything yet.

```ts
export interface DefaultedStateObservable <T> extends Observable<T> {
  getValue: () => T
}

// Example...
import { state } from '@rx-state/core';

const myState$ = state(interval(1000), 0);

const res = myState$.getValue();
expect(res).toBe(0);

myState$.subscribe(); // Start the subscription.

// Wait a couple of seconds

const res2 = myState$.getValue();
expect(res2).toBe(2);
```

**Comment by mattkime** [comment]
I came here to make this request. I'm glad to see other people making similar requests although I'm disappointed this hasn't moved forward. Here's my naive attempt at implementing this - `type BehaviorObservable<T> = Omit<BehaviorSubject<T>, 'next'>;` - it looks awfully close to the original proposal, enough so that I'm honestly not sure if there's a difference.
